go-controller/pkg/crd/egressfirewall/v1/types.go

```diff
@@ -0,0 +1,80 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EgressNetworkFirewallRuleType indicates whether an EgressNetworkFirewallRule allows or denies traffic
+// +kubebuilder:validation:Pattern=^Allow|Deny$
+type EgressFirewallRuleType string
+
+const (
+	EgressFirewallRuleAllow EgressFirewallRuleType = "Allow"
+	EgressFirewallRuleDeny  EgressFirewallRuleType = "Deny"
+)
+
+// +genclient
+// +resource:path=egressfirewall
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// EgressFirewall describes the current egress firewall for a Namespace.
+// Traffic from a pod to an IP address outside the cluster will be checked against
+// each EgressFirewallRule in the pod's namespace's EgressFirewall, in
+// order. If no rule matches (or no EgressFirewall is present) then the traffic
+// will be allowed by default.
+type EgressFirewall struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of EgressFirewall.
+	Spec EgressFirewallSpec `json:"spec"`
+	// Observed status of EgressFirewall
+	// +optional
+	Status EgressFirewallStatus `json:"status"`
+}
+
+// EgressFirewallStatus describes the lifecycle status of EgressFirewall.
+type EgressFirewallStatus struct {
+}
+
+// EgressFirewallSpec is a desired state description of EgressFirewall.
+type EgressFirewallSpec struct {
+	// a collection of egress firewall rule objects
+	Egress []EgressFirewallRule `json:"egress"`
+}
+
+// EgressFirewallRule is a single egressfirewall rule object
+type EgressFirewallRule struct {
+	// type marks this as an "Allow" or "Deny" rule
+	Type EgressFirewallRuleType `json:"type"`
+	// ports mark what ports and protocols for the rule to affect
+	Ports []EgressFirewallPort `json:"ports,omitempty"`
+	// to is one or more targes that traffic is allowed/denied to
+	To EgressFirewallDestination `json:"to"`
+}
+
+// EgressFirewallPort specifies the port to allow or deny traffic to
+type EgressFirewallPort struct {
+	// protocol (tcp, udp, sctp) that the traffic must match.
+	// +kubebuilder:validation:Pattern=^tcp|udp|sctp$
+	Protocol *string `json:"protocol,omitempty"`
+	// port that the traffic must match
+	// +optional
+	PortNum *int32 `json:"port"`
```

Comment:
And here it's missing a `omitempty`....if that really is needed?

Reply:
added omitempty